nntrainer/include/optimizer.h

```diff
@@ -201,36 +150,82 @@ class Optimizer {
    */
-  void read(std::ifstream &file);
+  virtual void read(std::ifstream &file);
 
   /**
    * @brief     Save Training optimizer paramters from file
    * @param[in] file output stream file
    */
-  void save(std::ofstream &file);
+  virtual void save(std::ofstream &file);
 
   /**
-   * @brief     get the base name for the layer
-   * @retval    base name of the layer
+   * @brief setProperty by PropertyType
+   * @note By passing empty string, this can validate if @a type is valid
+   * @param[in] type property type to be passed
+   * @param[in] value value to be passed, if empty string is passed, do nothing
+   * but throws error when @a type is invalid
+   * @exception exception::not_supported     when property type is not valid for
+   * the particular layer
+   * @exception std::invalid_argument invalid argument
    */
-  std::string getBaseName() { return "Optimizer"; };
+  virtual void setProperty(const PropertyType type,
+                           const std::string &value = "");
+
+  /**
+   * @brief     get the base name for the optimizer
+   * @retval    base name of the optimizer
+   */
+  virtual std::string getBaseName() = 0;
 
-private:
+  /**
+   * @brief     validate the optimizer
+   */
+  virtual void checkValidation();
+
+protected:
   /**
    * @brief Optimizer Type
    */
   OptType type;
 
   /**
-   * @brief Optimizer Hyper Parmeters
+   * @brief     get Learning Rate for the given iteration
+   * @param[in] iteration Iteration for the learning rate
+   * @retval    Learning rate
    */
-  OptParam popt;
+  virtual double getLearningRate(int iteration);
 
+  float learning_rate; /** learning rate */
+  float decay_rate;    /** decay rate for learning rate */
+  float decay_steps;   /** decay steps for learning rate */
+  bool continue_train; /** Continue training with previous tensors for adam */
+
+private:
   /**
-   * @brief Internal Tensors for adam Optimizer
+   * @brief     initialize optimizer. Initialize Weight if it is adam
+   * @param[in] params Weight list
+   * @param[in] num_weights size of the array
+   * @param[in] setTensor true if the layer need weight update.
+   *            Input Layer and Batch Normalization layer won't need it.
+   *            Therefore, it sets false.
+   * @retval #ML_ERROR_NONE Successful.
+   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
    */
-  std::vector<std::pair<Tensor, Tensor>> weight_mv;
+  virtual int initialize(std::shared_ptr<Weight> params,
+                         unsigned int num_weights, bool setTensor);
+
+  /**
+   * @brief     apply gradient to the given weight
+   * @param[in] weight Weight and gradient set to be updated
+   * @param[in] tensor_idx Idx of this tensor in the tensors list
+   * @param[in] num_weights size of the array
+   * @param[in] iteration nth epoch number
+   * @note weight which is called upon can be assumed to be trainable
+   */
+  virtual void apply_gradient(Weight &weight, int tensor_idx, double updated_lr,
```

Comment:
I think double and float are mixed for the learning rate. Could you check once up?

Reply:
`updated_lr` is intentionally double. Learning rate is handled as a double internally. However, for public view and the initial value are stored as a float but any operation or intermediate results of learning rate are performed as double.

Reply:
Alright